aws/resource_aws_cloudhsm2_cluster.go

```diff
@@ -0,0 +1,384 @@
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"manufacturer_hardware_certificate": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+
+			"security_group_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"cluster_state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsCloudHsm2ClusterImport(
+	d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	d.Set("cluster_id", d.Id())
+	return []*schema.ResourceData{d}, nil
+}
+
+func describeCloudHsm2Cluster(clusterId string, meta interface{}) (*cloudhsmv2.Cluster, error) {
+	conn := meta.(*AWSClient).cloudhsmv2conn
+
+	filters := []*string{&clusterId}
+	result := int64(1)
+	out, err := conn.DescribeClusters(&cloudhsmv2.DescribeClustersInput{
+		Filters: map[string][]*string{
+			"clusterIds": filters,
+		},
+		MaxResults: &result,
+	})
+	if err != nil {
+		log.Printf("[WARN] Error on retrieving CloudHSMv2 Cluster (%s) when waiting: %s", clusterId, err)
+		return nil, err
+	}
+
+	var cluster *cloudhsmv2.Cluster
+
+	for _, c := range out.Clusters {
+		if *c.ClusterId == clusterId {
+			cluster = c
+		}
+	}
+	return cluster, nil
+}
+
+func resourceAwsCloudHsm2ClusterRefreshFunc(
+	d *schema.ResourceData, meta interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		cluster, err := describeCloudHsm2Cluster(d.Id(), meta)
+
+		if cluster == nil {
+			return 42, "destroyed", nil
+		}
+
+		if cluster.State != nil {
+			log.Printf("[DEBUG] CloudHSMv2 Cluster status (%s): %s", d.Id(), *cluster.State)
+		}
+
+		return cluster, *cluster.State, err
+	}
+}
+
+func resourceAwsCloudHsm2ClusterCreate(d *schema.ResourceData, meta interface{}) error {
+	cloudhsm2 := meta.(*AWSClient).cloudhsmv2conn
+
+	subnetIdsSet := d.Get("subnet_ids").(*schema.Set)
+	subnetIds := make([]*string, subnetIdsSet.Len())
+	for i, subnetId := range subnetIdsSet.List() {
+		subnetIds[i] = aws.String(subnetId.(string))
+	}
+
+	input := &cloudhsmv2.CreateClusterInput{
+		HsmType:   aws.String(d.Get("hsm_type").(string)),
+		SubnetIds: subnetIds,
+	}
+
+	backupId := d.Get("backup_identifier").(string)
+	if len(backupId) != 0 {
+		input.SourceBackupId = aws.String(backupId)
+	}
+
+	log.Printf("[DEBUG] CloudHSMv2 Cluster create %s", input)
+
+	var output *cloudhsmv2.CreateClusterOutput
+
+	err := resource.Retry(180*time.Second, func() *resource.RetryError {
+		var err error
+		output, err = cloudhsm2.CreateCluster(input)
+		if err != nil {
+			if isAWSErr(err, cloudhsmv2.ErrCodeCloudHsmInternalFailureException, "request was rejected because of an AWS CloudHSM internal failure") {
+				log.Printf("[DEBUG] CloudHSMv2 Cluster re-try creating %s", input)
+				return resource.RetryableError(err)
+			}
+			return resource.NonRetryableError(err)
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
```

Comment:
We should return context about the error here, e.g. `return fmt.Errorf("error creating CloudHSMv2 Cluster: %s", err)`